tracer/src/Datadog.Trace.Coverage.collector/AssemblyProcessor.cs

```diff
@@ -0,0 +1,693 @@
+            {
+                return Instruction.Create(instruction.OpCode, instsOp);
+            }
+
+            if (instruction.Operand is VariableDefinition vDefOp)
+            {
+                return Instruction.Create(instruction.OpCode, vDefOp);
+            }
+
+            if (instruction.Operand is ParameterDefinition pDefOp)
+            {
+                return Instruction.Create(instruction.OpCode, pDefOp);
+            }
+
+            if (instruction.Operand is TypeReference tRefOp)
+            {
+                return Instruction.Create(instruction.OpCode, tRefOp);
+            }
+
+            if (instruction.Operand is float sOp)
+            {
+                return Instruction.Create(instruction.OpCode, sOp);
+            }
+
+            throw new Exception($"Instruction: {instruction.OpCode} cannot be cloned.");
+        }
+
+        private string CopyRequiredAssemblies(AssemblyDefinition assemblyDefinition, TracerTarget tracerTarget)
+        {
+            try
+            {
+                // Get the Datadog.Trace path
+                string targetFolder = "net461";
+                switch (tracerTarget)
+                {
+                    case TracerTarget.Net461:
+                        targetFolder = "net461";
+                        break;
+                    case TracerTarget.Netstandard20:
+                        targetFolder = "netstandard2.0";
+                        break;
+                    case TracerTarget.Netcoreapp31:
+                        targetFolder = "netcoreapp3.1";
+                        break;
+                    case TracerTarget.Net60:
+                        targetFolder = "net6.0";
+                        break;
+                }
+
+                var datadogTraceDllPath = Path.Combine(_tracerHome, targetFolder, "Datadog.Trace.dll");
+                var datadogTracePdbPath = Path.Combine(_tracerHome, targetFolder, "Datadog.Trace.pdb");
+
+                // Global lock for copying the Datadog.Trace assembly to the output folder
+                lock (PadLock)
+                {
+                    // Copying the Datadog.Trace assembly
+                    var assembly = typeof(Tracer).Assembly;
+                    var assemblyLocation = assembly.Location;
+                    var outputAssemblyDllLocation = Path.Combine(Path.GetDirectoryName(_assemblyFilePath) ?? string.Empty, Path.GetFileName(assemblyLocation));
+                    var outputAssemblyPdbLocation = Path.Combine(Path.GetDirectoryName(_assemblyFilePath) ?? string.Empty, Path.GetFileNameWithoutExtension(assemblyLocation) + ".pdb");
+                    if (!File.Exists(outputAssemblyDllLocation) ||
+                        assembly.GetName().Version >= AssemblyName.GetAssemblyName(outputAssemblyDllLocation).Version)
+                    {
+                        _logger.Debug($"GetTracerTarget: Writing {outputAssemblyDllLocation} ...");
```

Comment:
Could this cause any issues if they're referencing Datadog.Trace already? I see that we're overwriting if we have a newer version. I'm wondering about (contrived scenario) where we have 3 tracer version

- They reference Datadog.Trace directly (e.g. v2.7.0)
- They have the MSI installed (e.g. 2.9.0)
- They are using CI app (e.g. 2.10.0)

I can't think of any obvious issues, just thinking it through

Reply:
We should be safe in that scenario if we don't do breaking changes in the public api. The version mismatch should be handled like a normal production version mismatch scenario.

Reply:
> if we don't do breaking changes in the public api

Does this require that we have _both_ forward and back compatibility though 🤔 Currently we only have back compat (i.e. we can _add_ new public methods, but can't take them away). My concern is that this would require forward compat too (i.e. we can't add public methods). 

Struggling to get my head round if that's exactly the case though 🤔

Reply:
I cannot think of an scenario with forward compatibility issue. I kind of prefer to wait when this challenge comes.